gcore/gdalrasterband.cpp

```diff
@@ -1929,6 +1929,128 @@ uint64_t GDALRasterBand::GetNoDataValueAsUInt64(int *pbSuccess)
+            std::numeric_limits<uint8_t>::max());
+        if (dfNoDataValue == std::numeric_limits<unsigned char>::max())
+            dfReplacementVal = std::numeric_limits<unsigned char>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int8)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int8_t>::lowest(),
+            std::numeric_limits<int8_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt8>::max())
+            dfReplacementVal = std::numeric_limits<GInt8>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_UInt16)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<uint16_t>::lowest(),
+            std::numeric_limits<uint16_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GUInt16>::max())
+            dfReplacementVal = std::numeric_limits<GUInt16>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int16)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int16_t>::lowest(),
+            std::numeric_limits<int16_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt16>::max())
+            dfReplacementVal = std::numeric_limits<GInt16>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_UInt32)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<uint32_t>::lowest(),
+            std::numeric_limits<uint32_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GUInt32>::max())
+            dfReplacementVal = std::numeric_limits<GUInt32>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int32)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int32_t>::lowest(),
+            std::numeric_limits<int32_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt32>::max())
+            dfReplacementVal = std::numeric_limits<GInt32>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Float32)
+    {
+
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<float>::lowest(),
+            std::numeric_limits<float>::max());
+
+        const float fNoDataValue{static_cast<float>(dfNoDataValue)};
+
+        if (fNoDataValue == std::numeric_limits<float>::infinity())
+        {
+            dfNoDataValue = std::numeric_limits<float>::max();
+        }
+        else if (fNoDataValue == -std::numeric_limits<float>::infinity())
+        {
+            dfNoDataValue = std::numeric_limits<float>::lowest();
+        }
```

Comment:
Why? The infinity in case of float is the possible result of the down cast but in this case there is no cast that could generate and infinity.

Reply:
> The infinity in case of float is the possible result of the narrowing cast

not necessarily. A Float32 dataset could have a +inf nodata value (so not just because of a narrowing cast. And if we implement an early exit when the input dfNodataValue is out of range of float32, but maybe still allowing nan,inf,-inf, the narrowing cast shouldn't cause issues), and here your point was apparently to use a replacement value of <float>::max() if a computation would happen. I don't really think that in practice the result of a correct computation would lead to +inf, so my intuition would be that this particular case is not needed. But if it is needed for Float32, for the same reason, it should be needed for Float32.

Reply:
(sorry, my above comment was written before seeing your latest change. code review is hard :-))  Ok, so now we exit early if dfNoDataValue is +/- inf. So effectively the tests at line 2089-2090 could only trigger because of a narrowing cast, but it would seem weird to me that when dfNoDataValue <= std::numeric_limits<float>::max(),  static_cast<float>(dfNoDataValue) could be equal to std::numeric_limits<float>::infinity()